[WIP]AST transformer for isympy(IPython console)

ylemkimon opened the pull request:
- Implement AST transformers(IntegerWrapper and AutomaticSymbols) from Diofant(@skirpichev)
- Add automatic rationalization option: automatically wraps `float` with `Rational`, e.g., `0.5` to `Rational('0.5')`

* [ ] Implement token-based transformers or use IPython token-based transformer: #14440

- ~Change `optparse` to `argparse`: `optparse` is deprecated since Python version 2.7~
- ~Use existing IPython shell if present, when initializing new IPython session: Fixes #13418, fixes #13319, and fixes #11419.~
- ~Fixed incorrect architecture detection on 64-bit Windows. [(SO)](https://stackoverflow.com/q/1405913/1448742)~

Split into #14443 and #14440.

jksuom commented:
I think that IPython support should not be removed yet. SymPy works quite well with older versions.

asmeurer commented:
Also be aware of https://github.com/sympy/sympy/pull/13193

skirpichev commented:
> I think that IPython support should not be removed yet. SymPy works quite well with older versions.

Support for user-added AST transformations added, according to IPython log, in 1.0.0.

Any rational arguments for keeping compatibility with ancient IPython versions, where most isympy options will not work?

ylemkimon committed:
Fix architecture detection on 64-bit Windows

Fixed incorrect architecture detection on 64-bit Windows. See
https://stackoverflow.com/q/1405913/1448742.

ylemkimon committed:
Fix test failure with IPython < 2.0

IPython.core.formatters.FormatterWarning was introduced in IPython 2.0.

ylemkimon committed:
Add ast_transformers support for IPython < 1.0

Added ast_transformers support for IPython < 1.0 by monkey-patching
run_ast_nodes.

Commit:
Implement AST transformers from Diofant

Implemented AST transformers(IntegerWrapper and AutomaticSymbols) from
https://github.com/diofant/diofant/blob/master/diofant/interactive/session.py.

ylemkimon committed:
Change optparse to argparse in isympy

Changed optparse to argparse in isympy, since optparse is deprecated
since Python version 2.7.

ylemkimon committed:
Modify tests for IPython transformers

Modified tests for new IPython AST transformers:
* Removed readline dependecy
* Disabled history for test inputs
* Added int/int test

ylemkimon commented:
@jksuom Instead of removing IPython < 1.0 support, added ast_transformers support for IPython < 1.0 by monkey-patching run_ast_nodes.

ylemkimon commented:
Should we remove `sympy.interactive.session.int_to_Integer`, deprecate it, or leave as is?

ylemkimon committed:
Add automatic rationalization option to isympy

Added automatic rationalization option to isympy, which automatically
wraps float with Rational, e.g., 0.5 to Rational('0.5').

ylemkimon committed:
Use existing IPython shell when initializing IPython session

Use existing IPython shell if present, when initializing new IPython
session.

asmeurer commented:
I wonder if we can move the command line option help into the command line options, so that it doesn't show them twice. I'm assuming it was done the way it is now because of some optparse limitations.

asmeurer commented on sympy/interactive/session.py:
```diff
@@ -85,6 +90,76 @@ def _make_message(ipython=True, quiet=False, source=None):
+    def visit_Num(self, node):
+        if isinstance(node.n, float):
+            return ast.Call(func=ast.Name(id='Rational', ctx=ast.Load()),
+                            args=[ast.Str(s=repr(node.n))], keywords=[],
```
Is it possible to use the exact string form? If you enter a float that has more precision than a Python float, like `1.00000000000000000001`, it gets rounded before being wrapped.

skirpichev replied:
Not with ast.  It sees not a "string form", but a float literal.  Which is 1.0 in this case.

asmeurer replied:
I think you may be able to "find" the original float by tokenizing the input string and looking for the literal with the appropriate col_offset and lineno. I don't know if it will be robust, though.

asmeurer replied:
I'm not suggesting to tokenize - transform - untokenize (as we did before with Integer wrapping). That has issues, such as changing the code that's shown in tracebacks. You still use an ast transformer, but instead of using node.n, you use tokenize the input string, then match up the token that has the corresponding node.col_offset and lineno.  This is assuming the full original input string is available from the IPython API.

asmeurer replied:
For example:

```py
>>> ast.parse('2 + 1.0000000000000001').body[0].value.right.lineno
1
>>> ast.parse('2 + 1.0000000000000001').body[0].value.right.col_offset
4
>>> for token in tokenize.tokenize(io.BytesIO('2 + 1.0000000000000001'.encode('utf-8')).readline):
...     print(token)
TokenInfo(type=59 (BACKQUOTE), string='utf-8', start=(0, 0), end=(0, 0), line='')
TokenInfo(type=2 (NUMBER), string='2', start=(1, 0), end=(1, 1), line='2 + 1.0000000000000001')
TokenInfo(type=53 (OP), string='+', start=(1, 2), end=(1, 3), line='2 + 1.0000000000000001')
TokenInfo(type=2 (NUMBER), string='1.0000000000000001', start=(1, 4), end=(1, 22), line='2 + 1.0000000000000001')
TokenInfo(type=0 (ENDMARKER), string='', start=(2, 0), end=(2, 0), line='')
```

So (1, 4) corresponds to the token `TokenInfo(type=2 (NUMBER), string='1.000000000001', start=(1, 4), end=(1, 18), line='2 + 1.000000000001')`, which has the original string

ylemkimon replied:
@asmeurer The input string is not passed to `run_ast_nodes` nor available to AST transformers. We'd have to monkey-patch `run_cell`, but I'm afraid there is a lot of processing between `run_cell` call and AST transformation.

ylemkimon replied:
@asmeurer I'm not sure why showing transformed input in the traceback is a problem. IPython already does that for string-based transformers and line magics, and it may be more informative to users.

skirpichev replied:
> You still use an ast transformer, but instead of using node.n, you use tokenize the input string

This may work only if NodeTransformer will not shift original col_offset/lineno after modifications.

> We have to monkey-patch run_cell

No.  You just can use other available IPython transformations, string-based.  And parse & run ast transformation here, together with tokenization.

But IMHO, this way just overkill c.f. simple tokenize->fix->untokenize.

> I'm not suggesting to tokenize -> transform -> untokenize (as we did before with Integer wrapping). That has issues, such as changing the code that's shown in tracebacks.

I bet, it doesn't related to using tokenize, but to your monkey-patched run_cell() method.  You should use provided by IPython API, namely string-based input (registered in python_line_transforms).

asmeurer replied:
If a tokenize based transformer is no worse than an ast transformer then why are we changing to ast transformers in the first place?

skirpichev replied:
Good question for PR owner.  But lets make it clear: currently sympy doesn't use **any** IPython API for transformers.  Nor ast transformers, nor string-based.

ylemkimon replied:
@asmeurer As you said, IPython ast transformers do not show transformed input in the traceback–though I’m not sure it’s neccessary. This PR aims to use IPython APIs for better compatibility, since as @skirpichev said, it currently monkey-patches IPython.

ylemkimon replied:
Also, if we use string-based or token-based transformers, maybe we can hook existing transformers at `sympy.parsing.sympy_parser` like `convert_xor`.

asmeurer replied:
How does IPython handle its extensions of Python (like magics), which can appear inside of Python statements, like

```
for i in range(10):
    %time i + 1
```

Does IPython have a special Python parser, or is it simpler than that? 

I agree that an xor extension would also be useful, and it has to happen at tokenization, before the Python AST is created, due to they way ^ is parsed (it has a different precedence than **, so `x + y^2` would be `Xor(Add(x, y), 2)` in the AST).

ylemkimon replied:
@asmeurer IPython just processes line by line, and look for magic commands starting with percent signs, at the start of line and preprocesses it before passing to Python.

asmeurer commented:
We should probably do some manual testing for older versions of Python, and also for the notebook and qtconsole.

ylemkimon commented:
@asmeurer `argparse` still doesn't support argument help formatting. However, we can hide it by setting `help=argparse.SUPPRESS`. Also, tested it on IPython 0.13, 1.0, and latest version.

skirpichev commented:
> argparse still doesn't support argument help formatting.

What do you mean?  Hack with using module \__doc__'s - could be fully replaced by argparse.   Parameter description could be formatted arbitrarily complex (multiple lines, indentation, etc).

asmeurer commented:
Should the float -> rational flag also enable the integer wrapping flag? With just `-r`, we have

```py
In [1]: 1/2
Out[1]: 0.5

In [2]: 0.5
Out[2]: 1/2
```

ylemkimon commented:
@asmeurer I think so. If use cases overlap, we should consider combining both option as `-rationalize`, since it rationalizes floats and integer divisions.

ylemkimon commented:
I've opened an issue regarding TokenInputTransformer: https://github.com/ipython/ipython/issues/10893

asmeurer commented:
It looks like https://github.com/sympy/sympy/pull/14443 is a better PR to merge for now. Although I did test `-r` and it seems to work (it should be added to `-I` though).
